File: templates/ts/CallbacksChecker.ts
import { float32, int32, pointer, Finalizable } from "./types"
import { ResourceHolder } from "./ResourceManager"
import { Deserializer, deserializeAndCallCallback } from "%SERIALIZER_PATH%"
import { %NATIVE_MODULE_ACCESSOR% } from "%NATIVE_MODULE_PATH%"

enum CallbackEventKind {
    Event_CallCallback = 0,
    Event_HoldManagedResource = 1,
    Event_ReleaseManagedResource = 2,
}

const bufferSize = 1024
const buffer = new Uint8Array(bufferSize)
const deserializer = new Deserializer(buffer.buffer, bufferSize)
export function checkArkoalaCallbacks() {
    while (true) {
        let result = %NATIVE_MODULE_ACCESSOR%()._CheckArkoalaCallbackEvent(buffer, bufferSize)
        if (result == 0) break

        deserializer.resetCurrentPosition()
        const eventKind = deserializer.readInt32() as CallbackEventKind
        switch (eventKind) {
            case CallbackEventKind.Event_CallCallback: {
                deserializeAndCallCallback(deserializer)
                break;
            } 
            case CallbackEventKind.Event_HoldManagedResource: {
                const resourceId = deserializer.readInt32()
                ResourceHolder.instance().hold(resourceId)
                break;
            } 
            case CallbackEventKind.Event_ReleaseManagedResource: {
                const resourceId = deserializer.readInt32()
                ResourceHolder.instance().release(resourceId)
                break;
            }
            default: throw new Error(`Unknown callback event kind ${eventKind}`)
        }
    }
}

File: libohos/templates/kotlin/CallbacksChecker.kt
enum class CallbackEventKind(var value: Int) {
    Event_CallCallback(0),
    Event_HoldManagedResource(1),
    Event_ReleaseManagedResource(2)
}

private fun createDeserializer(bufferSize: Int): DeserializerBase {
    val buffer = InteropNativeModule._Malloc(bufferSize.toLong())
    return DeserializerBase(buffer, bufferSize)
}

val bufferSize: Int = 8 * 1024
val deserializer = createDeserializer(bufferSize)

public fun checkArkoalaCallbacks() {
    while (true) {
        val result = InteropNativeModule._CheckCallbackEvent(deserializer.asBuffer(), bufferSize)
        if (result == 0) {
            break
        }
        deserializer.resetCurrentPosition()
        val eventKind = deserializer.readInt32()
        if (eventKind == CallbackEventKind.Event_CallCallback.value) {
            deserializeAndCallCallback(deserializer)
        }
        else if (eventKind == CallbackEventKind.Event_HoldManagedResource.value) {
            val resourceId = deserializer.readInt32()
            ResourceHolder.hold(resourceId)
        }
        else if (eventKind == CallbackEventKind.Event_ReleaseManagedResource.value) {
            val resourceId = deserializer.readInt32()
            ResourceHolder.release(resourceId)
        }
        else { throw Exception("Unknown callback event kind ${eventKind}") }
    }
}
